test/performance/odp_ipsecfwd.c

```diff
@@ -0,0 +1,1277 @@
+		++stats->discards;
+		return false;
+	}
+
+	return true;
+}
+
+static inline void handle_new_packet(odp_packet_t pkt, odph_table_t fwd_tbl, stats_t *stats,
+				     int worker_id)
+{
+	pkt_handle_result_t res = PKT_OK;
+
+	if (odp_unlikely(odp_packet_has_error(pkt))) {
+		++stats->discards;
+		goto err;
+	}
+
+	if (odp_packet_has_udp(pkt))
+		res = handle_udp(pkt, stats);
+	else if (odp_packet_has_ipsec(pkt))
+		res = handle_ipsec(pkt, stats);
+
+	if (odp_unlikely(res == PKT_NOK))
+		goto err;
+
+	if (res == PKT_FWD) {
+		if (!forward_packet(pkt, fwd_tbl, stats, worker_id))
+			goto err;
+
+		++stats->fwd_pkts;
+	}
+
+	return;
+err:
+	odp_packet_free(pkt);
+}
+
+static inline odp_bool_t is_ipsec_in(odp_packet_t pkt)
+{
+	return odp_packet_user_ptr(pkt) == NULL;
+}
+
+static inline void handle_ipsec_packet(odp_packet_t pkt, odph_table_t fwd_tbl, stats_t *stats,
+				       int worker_id)
+{
+	odp_ipsec_packet_result_t result;
+	pkt_handle_result_t res = PKT_FWD;
+
+	if (odp_unlikely(odp_ipsec_result(&result, pkt))) {
+		is_ipsec_in(pkt) ? ++stats->ipsec_in_errs : ++stats->ipsec_out_errs;
+		goto err;
+	}
+
+	if (odp_unlikely(result.status.all != ODP_IPSEC_OK)) {
+		is_ipsec_in(pkt) ? ++stats->ipsec_in_errs : ++stats->ipsec_out_errs;
+		goto err;
+	}
+
+	if (is_ipsec_in(pkt) && odp_packet_has_udp(pkt))
+		res = handle_udp(pkt, stats);
+
+	if (odp_unlikely(res == PKT_NOK))
+		goto err;
+
+	if (res == PKT_FWD) {
+		if (!forward_packet(pkt, fwd_tbl, stats, worker_id))
+			goto err;
+
+		is_ipsec_in(pkt) ? ++stats->ipsec_in_pkts : ++stats->ipsec_out_pkts;
+	}
+
+	return;
+
+err:
+	odp_packet_free(pkt);
+}
+
+static inline void handle_ipsec_status_ev(odp_event_t ev, stats_t *stats)
+{
+	odp_ipsec_status_t status;
+
+	if (odp_unlikely(odp_ipsec_status(&status, ev) < 0 || status.result < 0))
+		++stats->status_errs;
+
+	odp_event_free(ev);
+}
+
+static void drain_events(void)
+{
+	odp_event_t ev;
+
+	while (true) {
+		ev = odp_schedule(NULL, ODP_SCHED_NO_WAIT);
+
+		if (ev == ODP_EVENT_INVALID)
+			break;
+
+		odp_event_free(ev);
+	}
+}
+
+static int handle_packets(void *args)
+{
+	thread_config_t *config = args;
+	odp_event_t evs[NUM_EVS];
+	int cnt, worker_id = odp_thread_id();
+	odp_event_t ev;
+	odp_event_type_t type;
+	odp_event_subtype_t subtype;
+	odp_packet_t pkt;
+	odph_table_t fwd_tbl = config->prog_config->fwd_tbl;
+	stats_t *stats = &config->stats;
+
+	odp_barrier_wait(&config->prog_config->init_barrier);
+
+	while (odp_atomic_load_u32(&is_running)) {
+		cnt = odp_schedule_multi_no_wait(NULL, evs, NUM_EVS);
+
+		if (cnt == 0)
+			continue;
+
+		for (int i = 0; i < cnt; ++i) {
+			ev = evs[i];
+			type = odp_event_types(ev, &subtype);
+			pkt = odp_packet_from_event(ev);
+
+			if (type == ODP_EVENT_PACKET) {
+				if (subtype == ODP_EVENT_PACKET_BASIC) {
+					handle_new_packet(pkt, fwd_tbl, stats, worker_id);
+				} else if (subtype == ODP_EVENT_PACKET_IPSEC) {
+					handle_ipsec_packet(pkt, fwd_tbl, stats, worker_id);
```

Comment:
Naming of packet processing functions is slightly confusing. For instance, handle_ipsec_packet() does not handle received IPsec packets but IPsec-processed packets (which can be IPsec or cleartext, depending on the direction). The 'handle' prefix is also not very informative.

I would suggest name changes along these lines:

handle_packets() -> worker_main()
handle_new_packet() -> packet_input()
handle_udp() -> udp_input()
handle_ipsec() -> ipsec_input()
handle_ipsec_packet() -> ipsec_completion()

Actually, there is a bit of duplication of logic between handle_new_packet() and handle_ipsec_packet() and it seems adding support for synchronous ipsec operations would be a bit difficult. This is rather subjective but I think might be clearer to organize the code so that handle_new_packet() is split to separate packet_input() and packet_output() functions, somehow like this:

```
worker_main()
        while (ev = schedule())
                if (ev type is plain packet)
                        packet_input()
                else if (ev type is ipsec completion)
                        ipsec_completion()

packet_input()
        if (sa = ipsec_sa_lookup())
                ipsec_input()
        else
                packet_output()

ipsec_input()
        if (async)
                ipsec_in_enqueue()
        else
                ipsec_in()
                packet_output()

packet_output()
        if (sa = ipsec_sa_lookup())
                ipsec_output()
        else
                packet_forward()

ipsec_output()
        if (async)
                ipsec_out_enqueue()
        else
                ipsec_out()
                packet_forward()

ipsec_completion()
        if (inbound packet)
                packet_output()
        else
                packet_forward()

```

What do you think?

Reply:
I refactored the call paths in v3, they do make a bit more sense now.

About the naming, I prefer having verbs in the function names, though "handle" is a bit lazy I agree so renamed them.